File: labs/module8/Fortran/host_data/Solution/main.f90
program main

   use dot_m
   use dot_cuda_m

   integer :: i, j, m, n
   integer :: AllocateStatus
   real, allocatable :: A(:,:)
   real, allocatable :: B(:,:)
   real, allocatable :: C(:)
   real, allocatable :: D(:)

   m = 10
   n = 10

   write(*,*) 'Running with m,n: ', m, n

   ! alloc arrays (A to D)
   allocate( A(m,n), stat = AllocateStatus)
   IF (AllocateStatus /= 0) STOP "*** Not enough memory ***"

   allocate( B(m,n), stat = AllocateStatus)
   IF (AllocateStatus /= 0) STOP "*** Not enough memory ***"

   allocate( C(m), stat = AllocateStatus)
   IF (AllocateStatus /= 0) STOP "*** Not enough memory ***"

   allocate( D(m), stat = AllocateStatus)
   IF (AllocateStatus /= 0) STOP "*** Not enough memory ***"

   ! Create random arrays A and B
   call random_seed()
   call random_number(A)
   call random_number(B)
   do j = 1,n 
      do i = 1,m
          A(i,j) = int(A(i,j)*10.0)
          B(i,j) = int(B(i,j)*10.0)
      end do
   end do

   ! call dot product on host
   call dot(A, B, C, m, n)
   
!$acc data copyin(A,B) copyout(D)

!$acc host_data use_device(A,B,D)
   call dot_cuda(A, B, D, m, n)
!$acc end host_data

!$acc end data 

   ! check answers
   write(*,*) 'maxval(abs(C-D)): ', maxval(abs(C-D))

end program

File: labs/module8/C/host_data/Solution/main.c
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>
#include <unistd.h>

extern void dot_cuda(int*, int*, int*, int, int);

int main()
{

	int i, j, m, n;
	int *A, *B, *C, *D;

	srand(0);

	m = 4098;
	n = 4098;

	A = (int*) malloc( m*n * sizeof(int));
	B = (int*) malloc( m*n * sizeof(int));
	C = (int*) malloc(  m  * sizeof(int));
	D = (int*) malloc(  m  * sizeof(int));

	for( i = 0; i < m; i++ ) {
		for( j = 0; j < n; j++ ) {
			A[i*n+j] = rand() % 100 + 1;
			B[i*n+j] = rand() % 100 + 1;
		}
	}

	dot(A, B, C, m, n);

#pragma acc data copyin(A[0:m*n],B[0:m*n]) copyout(D[0:m])
{
#pragma acc host_data use_device(A,B,D)
{
	dot_cuda(A, B, D, m, n);
}
}

	for( i = 0; i < m; i++ ) {
		if( C[i] != D[i]) {
			printf("Error at index %i\n", i);
			return 0;
		}
	}

	printf("Program finished sucessfully.\n");
	return 0;

}
